sklearn/preprocessing/data.py

```diff
@@ -410,32 +541,101 @@ def fit(self, X, y=None):
 
         Parameters
         ----------
-        X : array-like or CSR matrix with shape [n_samples, n_features]
+        X : array-like or CSR or CSC matrix with shape [n_samples, n_features]
             The data used to compute the mean and standard deviation
             used for later scaling along the features axis.
+
+        y: Passthrough for ``Pipeline`` compatibility.
         """
-        X = check_array(X, accept_sparse='csr', copy=self.copy,
+        X = check_array(X, accept_sparse=('csr', 'csc'), copy=self.copy,
                         ensure_2d=False, warn_on_dtype=True,
                         estimator=self, dtype=FLOAT_DTYPES)
+
         if sparse.issparse(X):
+            pass
             if self.with_mean:
                 raise ValueError(
                     "Cannot center sparse matrices: pass `with_mean=False` "
                     "instead. See docstring for motivation and alternatives.")
-            self.mean_ = None
+
+            self.mean_, self.var_ = mean_variance_axis(X, axis=0)
+            if not self.with_std:
+                self.var_ = None
+        else:
+            self.mean_ = X.mean(axis=0)
 
             if self.with_std:
-                var = mean_variance_axis(X, axis=0)[1]
-                self.std_ = np.sqrt(var)
-                self.std_ = _handle_zeros_in_scale(self.std_)
+                self.var_ = X.var(axis=0)
             else:
-                self.std_ = None
-            return self
+                self.var_ = None
+
+        if self.with_std:
+            self.std_ = np.sqrt(self.var_)
+            self.scale_ = _handle_zeros_in_scale(self.std_)
         else:
-            self.mean_, self.std_ = _mean_and_std(
-                X, axis=0, with_mean=self.with_mean, with_std=self.with_std)
+            self.std_, self.scale_ = None, None
+
+        self.n_samples_seen_ = X.shape[0]
+        return self
+
+    def partial_fit(self, X, y=None):
```

Comment:
it looks to me like all the logic of fit is duplicated in partial fit. Can't we write partial fit in a way that fit is a special case?